File: TrainingWithNodeJS/tetrisModule.js
const boardDimensions = {
	width: 10,
	height: 20
};
const tetrominoes = [
	[	/* cyan */
		[ 7, 7, 7, 7 ],
		[ 0, 0, 0, 0 ],
		[ 7, 7, 7, 7 ],
		[ 7, 7, 7, 7 ],
	],
	[	/* "blue" */
		[ 1, 7, 7 ],
		[ 1, 1, 1 ],
		[ 7, 7, 7 ],
	],
	[	/* "orange" */
		[ 7, 7, 2 ],
		[ 2, 2, 2 ],
		[ 7, 7, 7 ],
	],
	[	/* "yellow" */
		[ 3, 3 ], 
		[ 3, 3 ], 
	],
	[	/* "green" */
		[ 7, 4, 4 ],
		[ 4, 4, 7 ],
		[ 7, 7, 7 ],
	],
	[	/* "purple" */
		[ 7, 5, 7 ],
		[ 5, 5, 5 ],
		[ 7, 7, 7 ],
	],
	[	/* "red" */
		[ 6, 6, 7 ],
		[ 7, 6, 6 ],
		[ 7, 7, 7 ],
	]
];
const scoresForLines = [0, 40, 100, 300, 1200];
const perfectClearScore = [0,800, 1200, 1800, 2000, 3200];


function generateAPiece() {
	const pieceNum = Math.floor(Math.random() * tetrominoes.length);
	return {
		pieceNum: pieceNum,
		shape: [...tetrominoes[pieceNum]],
		position: {
			x: 3, y: 0
		},
		rotations: 0,
	}
}


function canPieceFit(piece, board) {
	
	// Loop through rows of piece at locations in board and check to see if either or both are 7
	for (let j = 0; j < piece.shape.length; j++) {

		for (let i = 0; i < piece.shape[j].length; i++) {

			if (piece.shape[j][i] != 7 && (piece.position.y + j >= boardDimensions.height || (board[piece.position.y + j][piece.position.x + i] != 7))) 
				return false;
		}
	}

	return true;
}

function movePieceDown(piece, board) {
	if (!canPieceFit(piece, board)) return false;
	piece.position.y++;
	if (canPieceFit(piece, board)) {
		return true;
	} else {
		piece.position.y--;
		return false;
	}
}

function rotatePiece(piece, direction) {

	const size = piece.shape.length
	let rotatedMatrix = []

	// Initialize the rotatedMatrix with zeros
	for (let i = 0; i < size; i++) {
		rotatedMatrix.push(new Array(size).fill(0));
	}

	if (direction === "right") {
		for (let j = 0; j < size; j++) {
			for (let i = 0; i < size; i++) {
				rotatedMatrix[i][size - 1 - j] = piece.shape[j][i];
			}
		}
	} else if (direction === "left") {
		for (let j = 0; j < size; j++) {
			for (let i = 0; i < size; i++) {
				rotatedMatrix[size - 1 - i][j] = piece.shape[j][i];
			}
		}
	}

	piece.shape = rotatedMatrix;
}

function findPieceShadow(piece, board) {
	const pieceShadow = structuredClone(piece);
	const boardShadow = structuredClone(board);

	let movedDown = 0;

	while (movePieceDown(pieceShadow, boardShadow)) {
		movedDown++;
	}
	pieceShadow.movedDown = movedDown;

	// Draw the piece onto the board
	for (let j = 0; j < pieceShadow.shape.length && pieceShadow.position.y + j < boardDimensions.height; j++) {

		for (let i = 0; i < pieceShadow.shape[j].length && pieceShadow.position.x + i < boardDimensions.width; i++) {

			if (pieceShadow.shape[j][i] != 7) {
				boardShadow[pieceShadow.position.y + j][pieceShadow.position.x + i] = pieceShadow.shape[j][i]
			}
		}
	}

	return [pieceShadow, boardShadow];
}

function findBoardScore(board, weightConfig, timeMovedDown) {

	let crackiness = 0;			// Spaces that are empty and in the column directly adjacent to the current one
	let unreachableHoles = 0;	// Holes which we can't ever fill. They are inaccessible
	let unfillableHoles = 0;	// Holes which we can only partially fill. 3 deep or deeper
	let boardClearScore = 0;
	let rightColumnCount = 0;
	let linesCleared = 0;

	// Count the number of blocks in the rightmost column
	// The row of the minimum depth block in a column
	let depthList = [boardDimensions.height, boardDimensions.height, boardDimensions.height, boardDimensions.height, boardDimensions.height, boardDimensions.height, boardDimensions.height, boardDimensions.height, boardDimensions.height, boardDimensions.height];
	let columnCounts = [0,0,0,0,0,0,0,0,0,0];	// The counts of blocks in each column
	let wellSize = [0,0,0,0,0,0,0,0,0,0];		// Used to keep track of the size of a well in each column


	// Yes, yes the following code might look a bit weird
	// I'm trying to avoid looping through columns in the board by just going through it line by line
	// It may not offer the speed benefits I think it does (given the scale), but its fun to think like this
	// and to practice!

	let rightCounter = 0;
	let leftCounter = 0;

	for (let j = 0; j < boardDimensions.height; j++) {
		let blocksInLine = 0;

		// Do an initial scan to update non-holes
		for (let i = 0; i < boardDimensions.width; i++) {
			// Update heights if need be
			if (board[j][i] != 7) {
				if (depthList[i] == boardDimensions.height) depthList[i] = j;
				
				// Increase column counts
				columnCounts[i]++;
				
				// Increase the number of blocks in the line
				blocksInLine++;

				if (wellSize[i] >= 3) {
					unfillableHoles += wellSize[i] ** weightConfig.wellCoefficient;
				}

				// Reset well counter
				wellSize[i] = 0;
			}

		}

		for (let i = 0; i < boardDimensions.width; i++) {
			if (board[j][i] == 7) {		// Holes

				// Adjust crackiness. It's just adding booleans, calm down Noah
				let crackCount = 0;

				leftCounter += (i != 0 && j >= depthList[i-1]);
				rightCounter += (i != 0 && j >= depthList[i+1]);

				crackCount += (i != 0 && j >= depthList[i-1]);	// If column is higher left of us
				crackCount += (i != 9 && j >= depthList[i+1]);	// If column is higher right of us
				
				crackiness += crackCount * (boardDimensions.height - j + 1) ** weightConfig.crackinessCoefficient;
				
				// Adjust unreachable holes
				if (j > depthList[i]) {
					unreachableHoles += (boardDimensions.height - j + 1) ** weightConfig.unreachableCoefficient;
				}

				// Increase well size
				if (j >= depthList[i]) wellSize[i]++;

			}
		}

		// Increment the right column counter if necessary
		if (board[j][boardDimensions.width - 1] != 7) rightColumnCount++;

		// Check to see if we cleared a line
		if (blocksInLine == boardDimensions.width) linesCleared++;
	}


	// Okay, after finding info, lets get some statistics
	// Find height aggregate
	let heightAgg = (boardDimensions.height * boardDimensions.width - depthList.reduce((a, b) => a + b)) / boardDimensions.width;

	// Now bumpiness
	let bumpiness = 0;
	for (let i = 1; i < boardDimensions.width; i++) {
		bumpiness += Math.abs(depthList[i] - depthList[i-1]);
	}

	// Find highest block
	let highestBlock = boardDimensions.height - depthList.reduce( (a, b) => a < b ? a : b)

	// Average the amount in columns
	let averageColumnLoad = columnCounts.reduce((a, b) => a + b) / boardDimensions.width;

	// Factor in the number of points the move would have given us
	let scoreValue = scoresForLines[linesCleared] ** weightConfig.scoreCoefficient;
	if (highestBlock - linesCleared == 0) boardClearScore = perfectClearScore[linesCleared] ** weightConfig.perfectScoreCoefficient;

	// console.log(depthList);

	// console.log("crackiness: " + crackiness);
	// console.log("unreachableHoles: " + unreachableHoles);
	// console.log("unfillableHoles: " + unfillableHoles);
	// console.log("heightAgg: " + heightAgg);
	// console.log("linesCleared: " + linesCleared);
	// console.log("bumpiness: " + bumpiness);
	// console.log("highestBlock: " + highestBlock);
	// console.log("averageColumnLoad: " + averageColumnLoad);
	// console.log("scoreValue: " + scoreValue);
	// console.log("rightColumnCount: " + rightColumnCount);
	// console.log("timeMovedDown: " + timeMovedDown);
	// console.log("boardClearScore: " + boardClearScore);

	let finalScore = weightConfig.crackiness * crackiness +
				weightConfig.unreachableHoles * unreachableHoles +
				weightConfig.unfillableHoles * unfillableHoles +
				weightConfig.aggregate * heightAgg +
				weightConfig.linesCleared * linesCleared +
				weightConfig.bumpiness * bumpiness +
				weightConfig.highestBlock * highestBlock +
				weightConfig.averageColumnLoad * averageColumnLoad +
				weightConfig.scoreValue * scoreValue +
				weightConfig.rightColumn * rightColumnCount +
				weightConfig.movedDown * timeMovedDown +
				weightConfig.boardClearScore * boardClearScore;

	return finalScore;
}

function tryAllOrientations(piece, board, weightConfig) {
	
	let pieceCopy = structuredClone(piece);
	const boardCopy = structuredClone(board);
	
	const possibilities = [];
	
	for (t = 0; t < 4; t++) {
		pieceCopy.position = { x: 0, y: 0}
		// Try to move it left, this is really just for the i piece
		pieceCopy.position.x--;
		if (!canPieceFit(pieceCopy, boardCopy)) pieceCopy.position.x++;
		pieceCopy.position.x--;
		if (!canPieceFit(pieceCopy, boardCopy)) pieceCopy.position.x++;
		while (canPieceFit(pieceCopy, boardCopy) || pieceCopy.position.x + pieceCopy.shape.length < boardDimensions.width) {
			if (canPieceFit(pieceCopy, boardCopy)) {
				const shadConfig = findPieceShadow(pieceCopy, boardCopy);
				const newBoard = shadConfig[1];
				let score

				// If we are at the top we want to know what the best position for the next piece is
				score = findBoardScore(newBoard, weightConfig, shadConfig[0]);
				
				possibilities.push({
					rotations: t,
					position: {...pieceCopy.position},
					board: newBoard,
					score: score
				});
			}
			
			pieceCopy.position.x++;
		}
		rotatePiece(pieceCopy, "right")
	}

	return possibilities;
}

function runTetrisGame(weightConfig, maxPieces=Infinity) {

	const boardState = [
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,],
		[ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,]
	];

	var lines = 0;
	var score = 0;
	var iterationCount = 0;

	var currentPiece = generateAPiece();
	// Move it down
	while (movePieceDown(currentPiece, boardState)) {}
	
	var nextPiece = generateAPiece();
	var heldPiece = generateAPiece();

	function switchHeldPiece() {
		const temp = heldPiece;
		heldPiece = currentPiece;
		currentPiece = temp;
		heldPiece.position.x = 3;
		heldPiece.position.y = 1;
	}

	function autoIncrement() {
	
		
		// Check to see if rows are filled
		let linesCleared = 0;
		for (let j = boardDimensions.height-1; j > 2; j--) {
			
			if (boardState[j].every(color => color != 7)) {
				// Level cleared, remove the row and then add a new one at the top

				boardState.splice(j, 1);
				boardState.unshift( [7, 7, 7, 7, 7, 7, 7, 7, 7, 7] );
				
				j++;
				lines++;
				linesCleared++;
			}

			
		}
		score += scoresForLines[linesCleared];

		
		let noneLeft = !boardState.some( row => row.some( block => block != 7));
		
		if (noneLeft) {
			if (linesCleared == 4) {
				if (prevRowTetrisClearOut) {
					// We got 2 tetris's in a row
					score += perfectClearScore[5];
				}
				prevRowTetrisClearOut = true;
			} else {
				prevRowTetrisClearOut = false;
				score += perfectClearScore[linesCleared];
			}
		} else {
			prevRowTetrisClearOut = false;
		}

		// Pick a new piece
		iterationCount++;
		
		currentPiece = {...nextPiece};
		nextPiece = generateAPiece();

		// Try to find best orientation
		const allOrientationsCurrent = tryAllOrientations(currentPiece, boardState, weightConfig);
		let currBest = Infinity;
		let currInd = 0;
		if (allOrientationsCurrent.length > 0) {
			for (let i = 0; i < allOrientationsCurrent.length; i++) {
				if (allOrientationsCurrent[i].score < currBest) {
					currInd = i;
					currBest = allOrientationsCurrent[i].score;
				}
			}
		}
		// Also try to find best orientation of held piece
		const allOrientationsHeld = tryAllOrientations(heldPiece, boardState, weightConfig);
		let heldBest = Infinity;
		let heldInd = 0;
		if (allOrientationsHeld.length > 0) {
			for (let i = 0; i < allOrientationsHeld.length; i++) {
				if (allOrientationsHeld[i].score < heldBest) {
					heldInd = i;
					heldBest = allOrientationsHeld[i].score;
				}
			}
		}

		
		if (allOrientationsCurrent.length == 0 && allOrientationsHeld.length == 0) {
			return false;
		}
		
		if (heldBest < currBest && allOrientationsHeld.length > 0) {
			for (let i = 0; i < boardState.length; i++) {
				boardState[i] = allOrientationsHeld[heldInd].board[i];
			}
			switchHeldPiece();
		} else if (allOrientationsCurrent.length > 0) {
			for (let i = 0; i < boardState.length; i++) {
				boardState[i] = allOrientationsCurrent[currInd].board[i];
			}
		} else {
			return false;
		}
	
	
		return true;
	
	}
	
	let running = true;
	
	while(running && iterationCount < maxPieces)
		running = autoIncrement();

	return {linesCleared: lines, score: score, percentSurvived: iterationCount / (maxPieces > 1000000000 ? 1 : maxPieces)};
}

module.exports = { runTetrisGame };

File: TrainingWithCpp/TetrisGame/tetrisModule.cpp
#include "tetrisModule.hpp"
#include <iostream>
#include <vector>
#include <numeric>
#include "cmath"
#include <algorithm>
#include "../TetrisGame/possibility.hpp"
#include "piece.hpp"


#include <chrono>
#include <thread>

const int scoresForLines[] 		= { 0, 40, 100, 300, 1200 };
const int perfectClearScores[] 	= { 0, 800, 1200, 1800, 2000, 3200 };

double findBoardScore(TetrisBoard &board, AI_Instance &weightConfig, int timeMovedDown) {

	double crackiness = 0;			// Spaces that are empty and in the column directly adjacent to the current one
	double unreachableHoles = 0;	// Holes which we can't ever fill. They are inaccessible
	double unfillableHoles = 0;		// Holes which we can only partially fill. 3 deep or deeper
	double boardClearScore = 0;
	int rightColumnCount = 0;
	int linesCleared = 0;

	// Count the number of blocks in the rightmost column

	int depthList[BOARD_WIDTH] = {};	// The row of the minimum depth block in a column
	std::fill_n(depthList, BOARD_WIDTH, BOARD_HEIGHT);
	int columnCounts[BOARD_WIDTH] = {};	// The counts of blocks in each column
	int wellSize[BOARD_WIDTH] = {};		// Used to keep track of the size of a well in each column


	// Yes, yes the following code might look a bit weird
	// I'm trying to avoid looping through columns in the board by just going through it line by line
	// It may not offer the speed benefits I think it does (given the scale), but its fun to think like this

	int rightCounter = 0;
	int leftCounter = 0;

	for (int j = 0; j < BOARD_HEIGHT; j++) {
		int blocksInLine = 0;

		// Do an initial scan to update non-holes
		for (int i = 0; i < BOARD_WIDTH; i++) {
			// Update heights if need be
			if (board[j][i] != 7) {
				if (depthList[i] == BOARD_HEIGHT) depthList[i] = j;
				
				// Increase column counts
				columnCounts[i]++;
				
				// Increase the number of blocks in the line
				blocksInLine++;

				if (wellSize[i] >= 3) {
					unfillableHoles += pow(wellSize[i], weightConfig.wellCoefficient);
				}

				// Reset well counter
				wellSize[i] = 0;
			}

		}

		for (int i = 0; i < BOARD_WIDTH; i++) {
			if (board[j][i] == 7) {		// Holes

				// Adjust crackiness. It's just adding booleans, calm down Noah
				short crackCount = 0;

				leftCounter += (i != 0 && j >= depthList[i-1]);
				rightCounter += (i != 0 && j >= depthList[i+1]);

				crackCount += (i != 0 && j >= depthList[i-1]);	// If column is higher left of us
				crackCount += (i != 9 && j >= depthList[i+1]);	// If column is higher right of us
				
				crackiness += crackCount * pow((BOARD_HEIGHT - j + 1), weightConfig.crackinessCoefficient);
				
				// Adjust unreachable holes
				if (j > depthList[i]) {
					unreachableHoles += pow(BOARD_HEIGHT - j + 1, weightConfig.unreachableCoefficient);
				}

				// Increase well size
				if (j >= depthList[i]) wellSize[i]++;

			}
		}

		// Increment the right column counter if necessary
		if (board[j][BOARD_WIDTH - 1] != 7) rightColumnCount++;

		// Check to see if we cleared a line
		if (blocksInLine == BOARD_WIDTH) linesCleared++;
	}


	// Okay, after finding info, lets get some statistics
	// Find height aggregate
	double heightAgg = (double) (BOARD_HEIGHT * BOARD_WIDTH - std::accumulate(depthList, depthList + BOARD_WIDTH, 0)) / BOARD_WIDTH;

	// Now bumpiness
	int bumpiness = 0;
	for (short i = 1; i < BOARD_WIDTH; i++) {
		bumpiness += abs(depthList[i] - depthList[i-1]);
	}

	// Find highest block
	int highestBlock = BOARD_HEIGHT - std::accumulate(depthList, depthList + BOARD_WIDTH, BOARD_HEIGHT, [](int a, int b) {
		return a < b ? a : b;
	});

	// Average the amount in columns
	double averageColumnLoad = std::accumulate(columnCounts, columnCounts + BOARD_WIDTH, 0) / BOARD_WIDTH;

	// Factor in the number of points the move would have given us
	double scoreValue = pow(scoresForLines[linesCleared], weightConfig.scoreCoefficient);
	if (highestBlock - linesCleared == 0) boardClearScore = pow(perfectClearScores[linesCleared], weightConfig.perfectScoreCoefficient);

	// std::cout << "crackiness: " << crackiness << std::endl;
	// std::cout << "unreachableHoles: " << unreachableHoles << std::endl;
	// std::cout << "unfillableHoles: " << unfillableHoles << std::endl;
	// std::cout << "heightAgg: " << heightAgg << std::endl;
	// std::cout << "linesCleared: " << linesCleared << std::endl;
	// std::cout << "bumpiness: " << bumpiness << std::endl;
	// std::cout << "highestBlock: " << highestBlock << std::endl;
	// std::cout << "averageColumnLoad: " << averageColumnLoad << std::endl;
	// std::cout << "scoreValue: " << scoreValue << std::endl;
	// std::cout << "rightColumnCount: " << rightColumnCount << std::endl;
	// std::cout << "timeMovedDown: " << timeMovedDown << std::endl;
	// std::cout << "boardClearScore: " << boardClearScore << std::endl;

	double finalScore = weightConfig.crackiness * crackiness +
				weightConfig.unreachableHoles * unreachableHoles +
				weightConfig.unfillableHoles * unfillableHoles +
				weightConfig.aggregate * heightAgg +
				weightConfig.linesCleared * linesCleared +
				weightConfig.bumpiness * bumpiness +
				weightConfig.highestBlock * highestBlock +
				weightConfig.averageColumnLoad * averageColumnLoad +
				weightConfig.scoreValue * scoreValue +
				weightConfig.rightColumn * rightColumnCount +
				weightConfig.movedDown * timeMovedDown +
				weightConfig.boardClearScore * boardClearScore;

	return finalScore;
}

std::vector<Possibility>* tryAllOrientations(Piece piece, TetrisBoard &board, AI_Instance &weightConfig) {

	// Create the return vector
	std::vector<Possibility>* possibilities = new std::vector<Possibility>;

	// Loop through each possible turn of the piece
	for (int t = 0; t < 4; t++) {
		piece.position[0] = -1;	// We're trying to move it to the left because there are gaps in the matrix
		piece.position[1] = 0;
		if (!piece.canPieceFit(board)) piece.position[0]++;

		// Try to move it left again, this is really only for the i piece
		piece.position[0]--;
		if (!piece.canPieceFit(board)) piece.position[0]++;
		
		// Keep trying until the piece won't fit AND we're at the edge of the board
		for (int a = 0; a < BOARD_WIDTH; a++) {

			if (piece.canPieceFit(board)) {

				// Create a copy of the board
				TetrisBoard boardCopy;
				for (int j = 0; j < BOARD_HEIGHT; j++) {
					for (int i = 0; i < BOARD_WIDTH; i++) {
						boardCopy[j][i] = board[j][i];
					}
				}

				// Cast the piece shadow onto the board
				const int timeMovedDown = piece.findPieceShadow(boardCopy);
				const double score = findBoardScore(boardCopy, weightConfig, timeMovedDown);

				// Write our results to the return vector
				possibilities->push_back(Possibility(
					boardCopy,
					score
				));
			}
			piece.position[0]++;
		}
		
		piece.rotatePiece(true);
	}

	return possibilities;
}

void pieceCleanup(int &lines, int &score, bool &prevRowTetrisClearOut, TetrisBoard &board) {

	// First check to see if any of the lines got cleared
	int linesCleared = 0;
	for (int j = BOARD_HEIGHT - 1; j > 1; j--) {

		bool lineHasSpace = false;

		for (int i = 0; i < BOARD_WIDTH; i++) {

			// If any element in the row is empty, continue
			if (board[j][i] == 7) {
				lineHasSpace = true;
				break;
			}
		}
		// If the line has a space, we don't score it
		if (lineHasSpace) continue;

		// If the line has no spaces, we need to move the board down
		for (int k = j; k > 0; k--) {
			for (int l = 0; l < BOARD_WIDTH; l++) {
				board[k][l] = board[k-1][l];
			}
		}

		// Set the topmost row to all 7's
		for (int l = 0; l < BOARD_WIDTH; l++) {
			board[0][l] = 7;
		}

		// Make sure we stay on the same row
		j++;

		// Increment appropriate stuff
		lines++;
		linesCleared++;
	}

	score += scoresForLines[linesCleared];

	bool clearBoard = true;
	for (int j = BOARD_HEIGHT - 1; j > 2; j--) {
		for (int i = 0; i < BOARD_WIDTH; i++) {
			if (board[j][i] != 7) {
				clearBoard = false;
				break;
			}
		}
		if (!clearBoard) break;
	}

	if (clearBoard) {
		if (linesCleared == 4) {
			if (prevRowTetrisClearOut) {
				// 2 tetris's in a row!
				score += perfectClearScores[5];
			} else {
				score += perfectClearScores[4];
			}
			prevRowTetrisClearOut = true;
		} else {
			prevRowTetrisClearOut = false;
			score += perfectClearScores[linesCleared];
		}
	} else {
		prevRowTetrisClearOut = false;
	}

}

void switchHeldPiece(Piece &currentPiece, Piece &heldPiece) {
	Piece temp = currentPiece;
	currentPiece = heldPiece;
	heldPiece = temp;
}

void runTetrisGame(AI_Instance weightConfig, int maxPieces, int (&results)[3], std::mt19937 &generator) {

	TetrisBoard boardState = {
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 },
		{ 7, 7, 7, 7, 7, 7, 7, 7, 7, 7 }
	};

	int lines = 0;
	int score = 0;
	int piecesPlaced = 0;
	bool prevRowTetrisClearOut = false;
	
	Piece currentPiece = Piece(generator);
	Piece nextPiece    = Piece(generator);
	Piece heldPiece    = Piece(generator);

	bool running = true;

	while (running && piecesPlaced < maxPieces) {

		piecesPlaced++;

		// Find the best orientation to put a piece in
		int currBestInd = -1;
		double currBestScore = 0;
		int heldBestInd = -1;
		double heldBestScore = 0;


		// Find the best orientation of the current piece
		std::vector<Possibility>* allOrientationsCurrent = tryAllOrientations(currentPiece, boardState, weightConfig);
		if (allOrientationsCurrent->size() > 0) {
			currBestScore = (*allOrientationsCurrent)[0].score;
			currBestInd = 0;
			for (int i = 0; i < allOrientationsCurrent->size(); i++) {
				if ((*allOrientationsCurrent)[i].score < currBestScore) {
					currBestInd = i;
					currBestScore = (*allOrientationsCurrent)[i].score;
				}
			}
		}

		// Of the held piece
		std::vector<Possibility>* allOrientationsHeld = tryAllOrientations(heldPiece, boardState, weightConfig);
		if (allOrientationsHeld->size() > 0) {
			heldBestScore = (*allOrientationsHeld)[0].score;
			heldBestInd = 0;
			for (int i = 0; i < allOrientationsHeld->size(); i++) {
				if ((*allOrientationsHeld)[i].score < heldBestScore) {
					heldBestInd = i;
					heldBestScore = (*allOrientationsHeld)[i].score;
				}
			}
		}

		
		// If there are no orientations available, consider the game lost
		if (allOrientationsCurrent->size() == 0 && allOrientationsHeld->size() == 0) {
			running = false;

			// Get rid of the lists that got returned
			delete allOrientationsCurrent;
			delete allOrientationsHeld;
			
			continue;
		}

		// Pick the overall best orientation and stamp it on the board
		if (heldBestScore < currBestScore && heldBestInd != -1) {
			
			for (int j = 0; j < BOARD_HEIGHT; j++) {
				for (int i = 0; i < BOARD_WIDTH; i++) {
					boardState[j][i] = (*allOrientationsHeld)[heldBestInd].board[j][i];
				}
			}

			switchHeldPiece(currentPiece, heldPiece);
		} else if (currBestInd != -1) {
			
			for (int j = 0; j < BOARD_HEIGHT; j++) {
				for (int i = 0; i < BOARD_WIDTH; i++) {
					boardState[j][i] = (*allOrientationsCurrent)[currBestInd].board[j][i];
				}
			}
		} else {
			running = false;
			
			// Get rid of the lists that got returned
			delete allOrientationsCurrent;
			delete allOrientationsHeld;

			continue;
			// Why would it do this? Idk
			// I had it in my old code
		}



		// Get rid of the lists that got returned
		delete allOrientationsCurrent;
		delete allOrientationsHeld;

		// std::cout << "check6" << std::endl;
		
		pieceCleanup(lines, score, prevRowTetrisClearOut, boardState);

		// Get a new current piece
		currentPiece = nextPiece;
		nextPiece = Piece(generator);

	}

	results[0] = score;
	results[1] = lines;
	results[2] = piecesPlaced;

	return;
}